File: apps/benchmarks/modules/bench-probe/ios/BenchProbeModule.swift
import ExpoModulesCore
import UIKit

public class BenchProbeModule: Module {
  public func definition() -> ModuleDefinition {
    Name("BenchProbe")

    // Synchronous by design: the sampler polls this from the JS thread while
    // inference runs on the worklet thread, and a promise hop per sample would
    // add more jitter than the read itself costs.
    Function("memoryFootprintBytes") { () -> Double in
      var info = task_vm_info_data_t()
      var count = mach_msg_type_number_t(
        MemoryLayout<task_vm_info_data_t>.size / MemoryLayout<natural_t>.size)

      let result = withUnsafeMutablePointer(to: &info) { pointer in
        pointer.withMemoryRebound(to: integer_t.self, capacity: Int(count)) { rebound in
          task_info(mach_task_self_, task_flavor_t(TASK_VM_INFO), rebound, &count)
        }
      }

      // phys_footprint is what jetsam measures an app against, so it is the
      // number that decides whether a model fits on a device.
      return result == KERN_SUCCESS ? Double(info.phys_footprint) : -1
    }

    // No public API exposes the malloc-zone total separately from the footprint
    // on iOS, so the two agree here. Android is where they diverge.
    Function("nativeHeapBytes") { () -> Double in
      var info = mach_task_basic_info()
      var count = mach_msg_type_number_t(
        MemoryLayout<mach_task_basic_info>.size / MemoryLayout<natural_t>.size)

      let result = withUnsafeMutablePointer(to: &info) { pointer in
        pointer.withMemoryRebound(to: integer_t.self, capacity: Int(count)) { rebound in
          task_info(mach_task_self_, task_flavor_t(MACH_TASK_BASIC_INFO), rebound, &count)
        }
      }

      return result == KERN_SUCCESS ? Double(info.resident_size) : -1
    }

    Function("deviceInfo") { () -> [String: Any] in
      var systemInfo = utsname()
      uname(&systemInfo)
      let identifier = withUnsafePointer(to: &systemInfo.machine) { pointer in
        pointer.withMemoryRebound(to: CChar.self, capacity: 1) { String(cString: $0) }
      }

      // On a simulator `utsname.machine` is the host Mac's architecture, so the
      // real device identity comes from the environment instead.
      let simulatorModel = ProcessInfo.processInfo.environment["SIMULATOR_MODEL_IDENTIFIER"]

      return [
        "model": simulatorModel ?? identifier,
        "osVersion": UIDevice.current.systemVersion,
        "soc": identifier,
        "isEmulator": simulatorModel != nil,
        "cpuCores": ProcessInfo.processInfo.processorCount,
        "totalMemoryBytes": Double(ProcessInfo.processInfo.physicalMemory),
      ]
    }
  }
}

File: apps/benchmarks/modules/bench-probe/android/src/main/java/expo/modules/benchprobe/BenchProbeModule.kt
package expo.modules.benchprobe

import android.app.ActivityManager
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import android.os.Build
import android.os.Debug
import android.os.PowerManager
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class BenchProbeModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("BenchProbe")

    // Synchronous by design: the sampler polls this from the JS thread while
    // inference runs on the worklet thread, and a promise hop per sample would
    // add more jitter than the read itself costs.
    //
    // Total PSS walks /proc/self/smaps, so it is the expensive one (single-digit
    // milliseconds). It is also the only counter that sees the pages a
    // memory-mapped .pte actually has resident, which is most of a model's
    // footprint — hence the sampler's default 100 ms interval rather than
    // something tighter.
    Function("memoryFootprintBytes") {
      val info = Debug.MemoryInfo()
      Debug.getMemoryInfo(info)
      info.totalPss.toDouble() * 1024.0
    }

    Function("nativeHeapBytes") {
      Debug.getNativeHeapAllocatedSize().toDouble()
    }

    // Thermal state decides whether two runs are comparable at all. Running a
    // suite immediately after another one made every raw-execute metric 9% to
    // 51% slower (median 22%) with no code change, purely because the phone had
    // not cooled down. Without this the harness reports that as a regression.
    //
    // `getCurrentThermalStatus` needs API 29; below that the battery
    // temperature is the only signal available.
    Function("thermalState") {
      val powerManager =
        appContext.reactContext?.getSystemService(Context.POWER_SERVICE) as? PowerManager
      val status =
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q && powerManager != null) {
          powerManager.currentThermalStatus
        } else {
          -1
        }

      // Sticky broadcast, so this returns immediately without registering.
      val battery = appContext.reactContext?.registerReceiver(
        null,
        IntentFilter(Intent.ACTION_BATTERY_CHANGED)
      )
      val tenthsOfCelsius = battery?.getIntExtra(BatteryManager.EXTRA_TEMPERATURE, -1) ?: -1

      mapOf(
        "status" to status,
        "statusName" to when (status) {
          0 -> "none"
          1 -> "light"
          2 -> "moderate"
          3 -> "severe"
          4 -> "critical"
          5 -> "emergency"
          6 -> "shutdown"
          else -> "unknown"
        },
        "batteryTemperatureC" to if (tenthsOfCelsius > 0) tenthsOfCelsius / 10.0 else -1.0
      )
    }

    Function("deviceInfo") {
      val activityManager =
        appContext.reactContext?.getSystemService(Context.ACTIVITY_SERVICE) as? ActivityManager
      val memoryInfo = ActivityManager.MemoryInfo().also { activityManager?.getMemoryInfo(it) }

      mapOf(
        "model" to Build.MODEL,
        "osVersion" to Build.VERSION.RELEASE,
        "soc" to if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) Build.SOC_MODEL else "",
        // Emulator images report a generic/google build fingerprint; the ABI
        // check alone would miss an arm64 emulator on an Apple Silicon host.
        "isEmulator" to (Build.FINGERPRINT.startsWith("generic") ||
          Build.FINGERPRINT.contains("vbox") ||
          Build.FINGERPRINT.contains("emulator") ||
          Build.MODEL.contains("Emulator") ||
          Build.MODEL.contains("Android SDK built for")),
        "cpuCores" to Runtime.getRuntime().availableProcessors(),
        "totalMemoryBytes" to memoryInfo.totalMem.toDouble()
      )
    }
  }
}
